sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/MapAggregate.scala

```diff
@@ -0,0 +1,332 @@
+      baseMap: mutable.HashMap[T, Long],
+      otherMap: mutable.HashMap[T, Long],
+      numBins: Int): Unit = {
+    otherMap.foreach { case (key, value) =>
+      if (baseMap.contains(key)) {
+        baseMap.update(key, baseMap(key) + value)
+      } else {
+        if (baseMap.size >= numBins) {
+          isInvalid = true
+          baseMap.clear()
+          return
+        } else {
+          baseMap.put(key, value)
+        }
+      }
+    }
+  }
+}
+
+// Digest class for column of string type.
+case class StringMapDigest(
+    bins: mutable.HashMap[UTF8String, Long] = mutable.HashMap.empty) extends MapDigest {
+  def this(bins: mutable.HashMap[UTF8String, Long], isInvalid: Boolean) = {
+    this(bins)
+    this.isInvalid = isInvalid
+  }
+
+  override def update(dataType: DataType, value: Any, numBins: Int): Unit = {
+    updateMap(bins, value.asInstanceOf[UTF8String], numBins)
+  }
+
+  override def merge(otherDigest: MapDigest, numBins: Int): Unit = {
+    mergeMap(baseMap = bins, otherMap = otherDigest.asInstanceOf[StringMapDigest].bins, numBins)
+  }
+
+  override def clear(): Unit = bins.clear()
+}
+
+object StringMapDigest {
+
+  private final def length(obj: StringMapDigest): Int = {
+    // isInvalid, size of bins
+    var len: Int = Ints.BYTES + Ints.BYTES
+    obj.bins.foreach { case (key, value) =>
+      // length of key, key, value
+      len += Ints.BYTES + key.getBytes.length + Longs.BYTES
+    }
+    len
+  }
+
+  final def serialize(obj: StringMapDigest): Array[Byte] = {
+    val buffer = ByteBuffer.wrap(new Array(length(obj)))
+    buffer.putInt(if (obj.isInvalid) 0 else 1)
+    buffer.putInt(obj.bins.size)
+    obj.bins.foreach { case (key, value) =>
+      val bytes = key.getBytes
+      buffer.putInt(bytes.length)
+      buffer.put(bytes)
+      buffer.putLong(value)
+    }
+    buffer.array()
+  }
+
+  final def deserialize(bytes: Array[Byte]): StringMapDigest = {
+    val buffer = ByteBuffer.wrap(bytes)
+    val isInvalid = if (buffer.getInt == 0) true else false
+    val size = buffer.getInt
+    val bins = new mutable.HashMap[UTF8String, Long]
+    var i = 0
+    while (i < size) {
+      val keyLength = buffer.getInt
+      var j = 0
+      val keyBytes = new Array[Byte](keyLength)
+      while (j < keyLength) {
+        keyBytes(j) = buffer.get()
+        j += 1
+      }
+      val value = buffer.getLong
+      bins.put(UTF8String.fromBytes(keyBytes), value)
+      i += 1
+    }
+    new StringMapDigest(bins, isInvalid)
+  }
+}
+
+// Digest class for column of numeric type.
+case class NumericMapDigest(
+    bins: mutable.HashMap[Double, Long] = mutable.HashMap.empty) extends MapDigest {
+  def this(bins: mutable.HashMap[Double, Long], isInvalid: Boolean) = {
+    this(bins)
+    this.isInvalid = isInvalid
+  }
+
+  override def update(dataType: DataType, value: Any, numBins: Int): Unit = {
+    // Use Double to represent endpoints (in histograms) for simplicity
+    val doubleValue = dataType match {
```

Comment:
Couldn't you leave this type promotion to the engine ?
That is, for numeric types instead of calling
`mapaggregate(numericcol, bins)`
call
`mapaggregate(cast(numericcol as double), bins)`

That way you only have to deal with doubles in your aggregate.